### PR TITLE
Drawing: Paths now have arrowheads, don't overlap and move.

### PR DESCRIPTION
**Pull request by jmatholia**
In experiment.js: On creation of path, paths now have arrow heads and the best edge is chosen so that there is no overlap of path and node. When node is moved all path's attached to it also move.

In CssGen.hs: svg graph now has a grid background, also required minor changes in experiment.js

Note: Elbows of paths do not move yet and if a node is moved the best edge is not recalculated, the path simply follows the node regardless of overlap. 

Also I think I should separate the functions in experiment.js since the file is getting quite long. Let me know if I should separate them before the "merge marathon".

### CODE REVIEW

File: js/experiment.js [hunk 1]
```diff
@@ -168,75 +187,183 @@ function nodeClicked(e) {
...
             curPath = null;
+            curElbow = null;
+        }
+    }
+}
+
+
+/*
+Return the best coordinates for the start and end of a edge. theNode and end could
+be a node or an elbow. Atleast one of beg and end must be a node.
```
Comment (Ian-Stewart-Binks) [inline js/experiment.js]: Atleast should be two words.

File: js/experiment.js [hunk 2]
```diff
@@ -168,75 +187,183 @@ function nodeClicked(e) {
...
+*/
+function findClosest(beg, typeB, end, typeE) {
+    'use-strict'
```
Comment (Ian-Stewart-Binks) [inline js/experiment.js]: it looks like a couple of functions are missing semi-colons after the `'use-strict'` statement.

File: js/experiment.js [hunk 3]
```diff
@@ -168,75 +187,183 @@ function nodeClicked(e) {
...
+
+    if (typeB === 'node' && typeE === 'elbow') {
+        var theNode = beg;
```
Comment (Ian-Stewart-Binks) [inline js/experiment.js]: I think that it is better to put duplicate declarations into one declaration at the top of the function, like `var theNode;`.

File: js/experiment.js (hunk 2 above)
Comment (Ian-Stewart-Binks) [inline js/experiment.js]: This should be `use strict;`. My mistake, I made a typo in the wiki. The typo is fixed now.

File: js/experiment.js [hunk 4]
```diff
@@ -126,7 +132,13 @@ function makeNode(e) {
                                     position.x > parseFloat(startNode.getAttribute('x')) + nodeWidth) &&
                     (position.y < startNode.getAttribute('y') || 
                                     position.y > parseFloat(startNode.getAttribute('y')) + nodeHeight)) {
-            curPath += 'L' + position.x + ',' + position.y + ' ';
+            if (curPath === null) {
+                findClosest({x: parseFloat(startNode.getAttribute('x')), y: parseFloat(startNode.getAttribute('y'))},
```
Comment (Ian-Stewart-Binks) [inline js/experiment.js]: Line's a bit long.

File: js/experiment.js [hunk 5]
```diff
@@ -109,10 +115,10 @@ function makeNode(e) {
         node.setAttribute('width', nodeWidth);
         node.setAttribute('height', nodeHeight);
         node.setAttribute('class', 'node');
-        node.attributes['parents'] = []; // node.parents = []
-        node.attributes['children'] = [];
-        node.attributes['inEdges'] = [];
-        node.attributes['outEdges'] = [];
+        /*node.parents = []; */ node.attributes['parents'] = [];
```
Comment (Ian-Stewart-Binks) [inline js/experiment.js]: Why did you move away from the dot notation?

File: js/experiment.js [hunk 6]
```diff
@@ -140,20 +152,27 @@ function nodeClicked(e) {
     var index = null;
 
     if (mode  === 'erase-mode') { 
-        // remove any paths leading to and from this node from the other node's list
+        // remove any paths leading to and from this node from the other node's list of paths
+        // and remove this node from the other nodes' adjacency lists
         e.currentTarget.attributes['inEdges'].map( function (item) {
```
Comment (Ian-Stewart-Binks) [inline js/experiment.js]: I don't think that there should be a space between `map(` and `function`.

**Comment by Ian-Stewart-Binks** [comment]
It looks like `nodeId` is a number, so the node ids probably start with numbers. In HTML5, ids need to start with alpha characters, this might be the reason why Firefox wasn't working in previous weeks.

**Comment by Ian-Stewart-Binks** [comment]
Lines 493-503 seem to have misaligned braces, and line 494 has extra spaces.

**Comment by Ian-Stewart-Binks** [comment]
In lines 7-12:
```
var colours = { 
     'red': '#D77546', 
     'green':'#2E8B57', 
     'blue':'#437699',
     'purple':'#46364A'
};
```

There should be spaces between all `:`, just like `'red'` has.

**Comment by Ian-Stewart-Binks** [comment]
Line 67 is missing a semi-colon.

File: hs/CssGen.hs [hunk 7]
```diff
@@ -639,7 +639,7 @@ aboutStyles = "#aboutDiv" ? do
     maxWidth (px 1000)
     padding 0 (em 1) 0 (em 1)
     margin nil auto nil auto
-    textAlign justify
+    textAlign justify
```
Comment (david-yz-liu) [inline hs/CssGen.hs]: Remove the extra space

File: hs/CssGen.hs [hunk 8]
```diff
@@ -673,13 +675,27 @@ drawStyles = do
...
-    width (pct 85)
+    width  (pct 85)
+    float floatRight
+    position relative
     "border-radius" -: "8px"
-    backgroundColor $ parse "#D3D3D3"
     border solid (px 2) "#000000"
 
+canvasBackground = "#background" ? do
+    height (pct 100)
```
Comment (david-yz-liu) [inline hs/CssGen.hs]: Use `height100` and `width100`

File: js/experiment.js [hunk 9]
```diff
@@ -168,75 +188,188 @@ function nodeClicked(e) {
...
     } else if (mode === 'path-mode') {
-        if (startNode === null) {
+        if (startNode === null || startNode === e.currentTarget) {
             // this is the start node of the path about to be created
             startNode = e.currentTarget;
-            curPath = 'M' + (parseFloat(startNode.getAttribute('x')) + nodeWidth/2) + 
-                                ',' + startNode.getAttribute('y') + ' ';   
-            console.log('starting it ' + curPath);
             select(e.currentTarget);
+            console.log('Path started ---------------');
```
Comment (david-yz-liu) [inline js/experiment.js]: Remove console.log

File: js/experiment.js [hunk 10]
```diff
@@ -168,75 +188,188 @@ function nodeClicked(e) {
...
+    if (typeB === 'node' && typeE === 'elbow') {
+        theNode = beg;
+        theElbow = end;
+    } else if (typeB === 'elbow' && typeE === 'node') {
+        // only need to add end point to curPath
+        theNode = end;
+        theElbow = beg;
+    } else {
+        // top, bottom, left, right
+        var node1Edges = [{x: beg.x + nodeWidth/2, y: beg.y},
```
Comment (david-yz-liu) [inline js/experiment.js]: Align the `{` for these.
